### PR TITLE
[Sparse] Remove the Python DiagMatrix class

### LINKED ISSUE #5367: [Sparse] Remove Python DiagMatrix class and represent it as a SparseMatrix.
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

There are two PRs for this issue.
1. Support C++ Diag sparse format. (#5432)
2. Remove DiagMatrix in Python and enable the Python sparse matrix with C++ Diag sparse matrix.

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description

Resolve #5367 .
1. Remove all DiagMatrix related code and use SparseMatrix for diagonal matrix computation.
2. Merge the tests of DiagMatrix into the test files of SparseMatrix.
3. The code of `elementwise_op_sp.py` and `elementwise_op.py` are not merged. They will be further refactored by issue #5412 .

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: python/dgl/sparse/elementwise_op.py [hunk 1]
```diff
@@ -2,43 +2,235 @@
...
-    +--------------+------------+--------------+--------+
-    |    A \\ B    | DiagMatrix | SparseMatrix | scalar |
-    +--------------+------------+--------------+--------+
-    |  DiagMatrix  |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    | SparseMatrix |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    |    scalar    |     🚫     |      🚫      |   🚫   |
-    +--------------+------------+--------------+--------+
+SparseMatrix.__add__ = sp_add
```
Comment (frozenbugs) [inline python/dgl/sparse/elementwise_op.py:212]: We do we need to duplicate each method?
What's the downside of SparseMatrix.__add__ = add?

In general the PR is oversize, in which 2 things are done.
One remove Python DiagMatrix, and use a new mechanism from the ops.
Consider breaking them to 2 PRs for easier to review.

Comment (czkkkkkk) [reply python/dgl/sparse/elementwise_op.py]: > We do we need to duplicate each method?
> What's the downside of SparseMatrix.add = add?

I didn't get your point. Do you mean why we don't define SparseMatrix.__add__ in sparse_matrix.py?

> Consider breaking them to 2 PRs for easier to review.

Done.

File: tests/python/pytorch/sparse/test_elementwise_op.py [hunk 2]
```diff
@@ -1,5 +1,4 @@
 import operator
-import sys
 
 import backend as F
```
Comment (frozenbugs) [inline tests/python/pytorch/sparse/test_elementwise_op.py:3]: Move the diag related elementwise ops test cases to this function?

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/sparse/elementwise_op.py (hunk 1 above)
Comment (frozenbugs) [reply python/dgl/sparse/elementwise_op.py]: The first comment is not related to this PR anymore.

File: python/dgl/sparse/elementwise_op.py [hunk 3]
```diff
@@ -2,43 +2,26 @@
...
-
-    +--------------+------------+--------------+--------+
-    |    A \\ B    | DiagMatrix | SparseMatrix | scalar |
-    +--------------+------------+--------------+--------+
-    |  DiagMatrix  |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    | SparseMatrix |     ✅     |      ✅      |   🚫   |
-    +--------------+------------+--------------+--------+
-    |    scalar    |     🚫     |      🚫      |   🚫   |
-    +--------------+------------+--------------+--------+
```
Comment (jermainewang) [inline python/dgl/sparse/elementwise_op.py:28]: Shall we keep the table to show that adding a sparse matrix with a scalar is not allowed?

File: python/dgl/sparse/elementwise_op_sp.py [hunk 4]
```diff
@@ -150,6 +163,12 @@ def sp_div(A: SparseMatrix, B: Scalar) -> SparseMatrix:
     """
     if is_scalar(B):
         return val_like(A, A.val / B)
+    if A.is_diag() and B.is_diag():
+        assert A.shape == B.shape, (
+            f"The shape of diagonal matrix A {A.shape} and B {B.shape} must"
+            f"match for elementwise division."
```
Comment (jermainewang) [inline python/dgl/sparse/elementwise_op_sp.py:169]: @frozenbugs I wonder if we will have some guidance on how to write error messages. Shape checks are particularly common.

**Review by jermainewang — APPROVED** [review]
Mostly fine. The only suggestion is to add back the table in the docstring. And I'll suggest @frozenbugs to update the quickstart tutorial to double check the correctness.

File: python/dgl/sparse/elementwise_op.py (hunk 3 above)
Comment (czkkkkkk) [reply python/dgl/sparse/elementwise_op.py:28]: I removed the tables because they are too simple. They are 2x2 and mostly with checkmarks.

**Comment by frozenbugs** [comment]
> Mostly fine. The only suggestion is to add back the table in the docstring. And I'll suggest @frozenbugs to update the quickstart tutorial to double check the correctness.

Yes, I will update the tutorial in the pre-release phrase.

File: python/dgl/sparse/elementwise_op_sp.py (hunk 4 above)
Comment (frozenbugs) [reply python/dgl/sparse/elementwise_op_sp.py:169]: I am fine if we just use the same pattern, in the future code review, we should just ask ppl to follow this patter.

It is probably not worth to create a shared function / string for this.

File: tests/python/pytorch/sparse/test_unary_op.py [hunk 5]
```diff
@@ -3,10 +3,24 @@
 import backend as F
 import torch
 
-from dgl.sparse import diag
+from dgl.sparse import diag, from_coo
```
Comment (frozenbugs) [inline tests/python/pytorch/sparse/test_unary_op.py:6]: instead using from_coo, we should use the canonical initiator. spmatrix.

**Comment by frozenbugs** [comment]
@dgl-bot

**Pull request merged**